execution_engine/src/storage/global_state/in_memory.rs

```diff
@@ -239,11 +245,26 @@ impl StateProvider for InMemoryGlobalState {
         &self,
         _correlation_id: CorrelationId,
         trie_key: &Digest,
-    ) -> Result<Option<Trie<Key, StoredValue>>, Self::Error> {
+        index: u64,
+    ) -> Result<Option<TrieOrChunkedData>, Self::Error> {
         let txn = self.environment.create_read_txn()?;
-        let ret: Option<Trie<Key, StoredValue>> = self.trie_store.get(&txn, trie_key)?;
-        txn.commit()?;
-        Ok(ret)
+        let bytes =
+            Store::<Digest, Trie<Digest, StoredValue>>::get_raw(&*self.trie_store, &txn, trie_key)?;
+        txn.commit()?; // TODO[RC]: Needed for read only txn?
+
+        bytes.map_or_else(
+            || Ok(None),
+            |bytes| {
+                if bytes.len() <= ChunkWithProof::CHUNK_SIZE_BYTES {
+                    let deserialized_trie = bytesrepr::deserialize(bytes.into())?;
+                    Ok(Some(TrieOrChunkedData::Trie(Box::new(deserialized_trie))))
+                } else {
+                    let chunk_with_proof = ChunkWithProof::new(bytes.as_slice(), index)
+                        .map_err(|_| error::Error::ChunkWithProofError)?;
+                    Ok(Some(TrieOrChunkedData::ChunkWithProof(chunk_with_proof)))
+                }
+            },
+        )
```

Comment:
Looks like we can't actually do that, since the code that would go into the closure uses the `?` operator. It would get a bit complex to get rid of that, and I think it's probably not worth it.